src/shims/fs.rs

```diff
@@ -722,6 +799,248 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         this.try_unwrap_io_result(result)
     }
+
+    fn mkdir(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        mode_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("mkdir")?;
+
+        let _mode = if this.tcx.sess.target.target.target_os.to_lowercase() == "macos" {
+            this.read_scalar(mode_op)?.not_undef()?.to_u16()? as u32
+        } else {
+            this.read_scalar(mode_op)?.to_u32()?
+        };
+
+        let path = this.read_os_str_from_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+
+        let mut builder = DirBuilder::new();
+
+        // If the host supports it, forward on the mode of the directory
+        // (i.e. permission bits and the sticky bit)
+        #[cfg(target_family = "unix")]
+        {
+            use std::os::unix::fs::DirBuilderExt;
+            builder.mode(_mode.into());
+        }
+
+        let result = builder.create(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn rmdir(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("rmdir")?;
+
+        let path = this.read_os_str_from_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+
+        let result = remove_dir(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn opendir(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("opendir")?;
+
+        let name = this.read_os_str_from_c_str(this.read_scalar(name_op)?.not_undef()?)?;
+
+        let result = read_dir(name);
+
+        match result {
+            Ok(dir_iter) => {
+                let id = this.machine.dir_handler.insert_new(dir_iter);
+
+                // The libc API for opendir says that this method returns a pointer to an opaque
+                // structure, but we are returning an ID number. Thus, pass it as a scalar of
+                // pointer width.
+                Ok(Scalar::from_int(id, this.pointer_size()))
+            }
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                Ok(Scalar::from_int(0, this.memory.pointer_size()))
+            }
+        }
+    }
+
+    fn readdir64_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir64_r")?;
+
+        let dirp = this.read_scalar(dirp_op)?.to_machine_usize(this)?;
+
+        let entry_ptr = this.force_ptr(this.read_scalar(entry_op)?.not_undef()?)?;
+        let dirent64_layout = this.libc_ty_layout("dirent64")?;
```

Comment:
I think these are only used inside the `match`; if yes, please move them there.

Reply:
Also maybe copy the definition of `dirent64` here, I will now have to find the libc source code to figure out what is going on here.